src/run/disjoint_set.h

```diff
@@ -0,0 +1,216 @@
+#include <algorithm>
+#include <stdio.h>
+#include <fstream>
+#include <limits>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+using namespace std;
+
+template<typename T> class Node {
+	Node<T>* parent = this;
+	uint32_t r = 0;
+public:
+	virtual ~Node(){};
+	virtual Node<T>* getParent(){
+		return parent;
+	}
+	virtual void setParent(Node<T>* p){
+		parent = p;
+	}
+
+	virtual const T* getValue() = 0;
+
+	virtual uint32_t getCount(){
+		return r;
+	}
+	virtual void incrementCount(){
+		r++;
+	}
+};
+
+template<typename T> class IntegralNode: public Node<T> {
+private:
+	T value;
+public:
+	virtual ~IntegralNode(){};
+	IntegralNode<T>(T v){
+		value = v;
+	}
+	virtual const T* getValue(){
+		return &value;
+	}
+};
+
+template<typename T> class TypeNode: public Node<T> {
+private:
+	const T* value;
+	uint32_t r;
+public:
+	virtual ~TypeNode(){};
+	TypeNode<T>(const T* v){
+		value = v;
+	}
+	virtual const T* getValue(){
+		return value;
+	}
+};
+
+template<typename T> class LazyDisjointSet {
+private:
+	virtual Node<T>* get(const T i) = 0;
+	virtual Node<T>* get(const T* i) = 0;
+	virtual uint64_t size() = 0;
+	virtual vector<Node<T>*>* getNodes() = 0;
+public:
+	virtual ~LazyDisjointSet(){};
+
+	virtual Node<T>* getRoot(Node<T>* x) {
+		if (x->getParent() != x) {
+			// flatten the tree while traversing
+			x->setParent(getRoot(x->getParent()));
+		}
+		return x->getParent();
+	}
+
+	virtual void merge(Node<T>* x, Node<T>* y) {
+		if (x != y) {
+			Node<T>* rootX = getRoot(x);
+			Node<T>* rootY = getRoot(y);
+			if (rootX != rootY) {
+				if (rootX->getCount() < rootY->getCount()) {
+					rootX->setParent(rootY);
+				}
+				else if (rootX->getCount() > rootY->getCount()) {
+					rootY->setParent(rootX);
+				}
+				else {
+					rootX->incrementCount();
+					rootY->setParent(rootX);
+				}
+			}
+		}
+	}
+
+	virtual void merge(const T* x, const T* y) {
+		merge(get(x), get(y));
+	}
+
+	void merge(T x, T y) {
+		merge(get(x), get(y)); 
+	}
+
+	virtual vector<unordered_set<T>> getListOfSets() {
+		unordered_map<const T*, unordered_set<T>*> map;
+		for (Node<T>* n : *(getNodes())) {
+			if (n == nullptr) {
+				continue;
+			}
+			const T* r = getRoot(n)->getValue();
+			auto it = map.find(r);
+			T v = *(n->getValue());
+			if (it == map.end()) {
+				it = map.emplace(r, new unordered_set<T>()).first;
+			}
+			it->second->emplace(v);
+		}
+		vector<unordered_set<T>> listOfSets;
+		for(auto el=map.begin(); el != map.end(); el++){
+			listOfSets.emplace_back(move(*(el->second)));
+		}
+		return listOfSets;
+	}
+};
+
+template<typename T> class LazyDisjointIntegralSet : public LazyDisjointSet<T> {
+private:
+	vector<Node<T>*> nodes;
```

Comment:
This would maybe be more efficient using a `vector<int>` to store the parent index for each sequence, to avoid memory allocations.

Reply:
I don't think this is the most expensive step right now. I will keep it in mind and return to it if it turns out to be a problem.